modules/host/negotiaterevisecontract.go

```diff
@@ -0,0 +1,337 @@
+			bandwidthRevenue = bandwidthRevenue.Add(settings.MinimumUploadBandwidthPrice.Mul(types.NewCurrency64(modules.SectorSize)))
+			storageRevenue = storageRevenue.Add(settings.MinimumStoragePrice.Mul(blockBytesCurrency))
+			collateralRisked = collateralRisked.Add(settings.Collateral.Mul(blockBytesCurrency))
+
+			// Insert the sector into the root list.
+			newRoot := crypto.MerkleRoot(sector[:])
+			sectorsGained = append(sectorsGained, newRoot)
+			gainedSectorData = append(gainedSectorData, sector[:])
+			so.SectorRoots = append(so.SectorRoots[:index], append([]crypto.Hash{newRoot}, so.SectorRoots[index:]...)...)
+		} else if modificationType == modules.ActionModify {
+			// Download the sector.
+			var offset uint64
+			var length uint64
+			err = encoding.ReadObject(conn, &offset, 8)
+			if err != nil {
+				return false, err
+			}
+			err = encoding.ReadObject(conn, &length, 8)
+			if err != nil {
+				return false, err
+			}
+			// Have to check all three cases, otherwise an attacker could abuse
+			// overflows.
+			if offset > modules.SectorSize || length > modules.SectorSize || offset+length > modules.SectorSize {
+				return false, errIllegalOffsetAndLength
+			}
+			var sectorModifications []byte
+			err = encoding.ReadObject(conn, &sectorModifications, length+8)
+			if err != nil {
+				return false, err
+			}
+
+			// Get the data for the new sector.
+			sector, err := h.readSector(so.SectorRoots[index])
+			if err != nil {
+				return false, err
+			}
+			for i := offset; i < uint64(len(sectorModifications)); i++ {
+				sector[i] = sectorModifications[i-offset]
+			}
+
+			// Update finances.
+			bandwidthRevenue = bandwidthRevenue.Add(settings.MinimumUploadBandwidthPrice.Mul(types.NewCurrency64(modules.SectorSize)))
+
+			// Update the sectors removed and gained to indicate that the old
+			// sector has been replaced with a new sector.
+			newRoot := crypto.MerkleRoot(sector[:])
+			sectorsRemoved = append(sectorsRemoved, so.SectorRoots[index])
+			sectorsGained = append(sectorsGained, newRoot)
+			gainedSectorData = append(gainedSectorData, sector[:])
+			so.SectorRoots[index] = newRoot
+		} else {
+			return false, errUnknownModification
+		}
+	}
+
+	// Read the file contract revision and check whether it's acceptable.
+	var revision types.FileContractRevision
+	err = encoding.ReadObject(conn, &revision, 16e3)
+	if err != nil {
+		return false, err
+	}
+	err = verifyRevision(so, revision, storageRevenue, bandwidthRevenue, collateralRisked)
+	if err != nil {
+		return false, rejectNegotiation(conn, err)
+	}
+
+	// Revision is acceptable, write an acceptance string.
+	err = encoding.WriteObject(conn, modules.AcceptResponse)
+	if err != nil {
+		return false, err
+	}
+
+	// Renter will now send the transaction signatures for the file contract,
+	// followed by an indication of whether another iteration is preferred.
+	var renterSig types.TransactionSignature
+	var another bool
```

Comment:
![thor](http://33.media.tumblr.com/tumblr_m728tfNO5t1rsy50k.gif)

This is a little confusing though lol. Maybe `shouldContinue` or `renterWantsContinue`?